### PR TITLE
Allow running with local conda environment and code.

### PR DESCRIPTION
**Pull request by kurtschelfthout**

### CODE REVIEW

File: src/meadowrun/deployment_manager.py [hunk 1]
```diff
@@ -222,8 +236,32 @@ async def _get_git_code_paths(
         return [os.path.join(local_copy_path, path_to_source)], local_copy_path
 
 
+async def _get_zip_file_code_paths(
+    local_copies_folder: str, zip_file_url: str, code_paths: Sequence[str]
+) -> List[str]:
+    decoded_url = urllib.parse.urlparse(zip_file_url)
```
Comment (hrichardlee) [inline src/meadowrun/deployment_manager.py:242]: We should probably add some local caching on the machine here? As it is, do you know what happens if the local `folder` already exists?

File: src/meadowrun/local_code.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    paths_to_zip = _get_paths_to_zip(additional_paths, include_sys_path)
+    if not paths_to_zip:
+        raise ValueError("No paths to zip")
+
+    real_paths_to_zip_paths, zip_paths = _consolidate_paths_to_zip(paths_to_zip)
+
+    # Currently we always zip all the files and the compute the hash.
+    # The hash is just used to avoid upload to S3. In future we could
+    # try to
```
Comment (hrichardlee) [inline src/meadowrun/local_code.py:24]: In the future we could try to???

File: src/meadowrun/run_job.py [hunk 3]
```diff
@@ -83,7 +89,49 @@ class Deployment:
...
+    async def mirror_local(
+        cls,
+        include_sys_path: bool = True,
+        additional_paths: Iterable[str] = tuple(),
+        conda_env: Optional[str] = None,
+        environment_variables: Optional[Dict[str, str]] = None,
+    ) -> Deployment:
+        """A deployment that mirrors local environment and code.
+
+        :param path_to_source: local code to mirror on the remote servers. Defaults to
```
Comment (hrichardlee) [inline src/meadowrun/run_job.py:106]: Looks like docstring is out of date

File: src/meadowrun/run_job.py [hunk 4]
```diff
@@ -599,6 +707,8 @@ async def run_map(
             )
             _add_deployments_to_job(job, code, interpreter)
 
+            # hmmm. prepare_job shouldn't run for every of these jobs...
```
Comment (hrichardlee) [inline src/meadowrun/run_job.py:710]: out of date comment?

File: src/meadowrun/run_job_local.py [hunk 5]
```diff
@@ -672,13 +672,22 @@ def _set_up_working_folder(
     git_repos_folder = os.path.join(working_folder, "git_repos")
     # see CodeDeploymentManager
     local_copies_folder = os.path.join(working_folder, "local_copies")
+    # misc folder for e.g. storing environment export files sent from local machine
+    misc_folder = os.path.join(working_folder, "misc")
```
Comment (hrichardlee) [inline src/meadowrun/run_job_local.py:676]: You should try to clean these up in SshHost.run_job in the finally clause

Comment (hrichardlee) [reply src/meadowrun/run_job_local.py:676]: Maybe...

**Review by hrichardlee — APPROVED** [review]

File: src/meadowrun/local_code.py (hunk 2 above)
Comment (kurtschelfthout) [reply src/meadowrun/local_code.py:24]: no one can know the future lol

**Pull request merged**